src/utils/update/review.beneficial.owner.ts

```diff
@@ -0,0 +1,106 @@
+  return boi?.usual_residential_address ? true : false;
+};
+
+// this check is to determine whether the BO has been fully submitted
+// by checking for appData submitted with form and not present after PSC fetch
+const checkBoOtherValidation = (boOther: BeneficialOwnerOther): boolean => {
+  return boOther?.principal_address ? true : false;
+};
+
+const checkForBackButtonBo = (appData: ApplicationData, boType: string, boRedirectUrl: string) => {
+  const boLength: number = appData[boType]?.length || 0;
+  const boIndex = boLength - 1;
+  const isAppDataAndBoLength = (appData[boType] && boLength >= 1);
+
+  if (
+    isAppDataAndBoLength && (
+      ((boType === AllBoTypes.boIndividual) && !checkBoIndividualValidation(appData[boType][boIndex]))
+    ||
+      ((boType === AllBoTypes.boOther) && !checkBoOtherValidation(appData[boType][boIndex]))
+    )
+  ) {
+    return `${boRedirectUrl}${boIndex}`;
+  }
+};
+
+export const checkAndReviewBeneficialOwner = (appData: ApplicationData): string => {
+  let redirectUrl = "";
+
+  for (const updateAppData in appData.update){
+    switch (updateAppData){
+        case AllBoTypes.boiReview: {
+          const boiFromBackButton = checkForBackButtonBo(appData, AllBoTypes.boIndividual, beneficialOwnerIndividualReviewRedirectUrl);
+          if (boiFromBackButton) {
+            redirectUrl = boiFromBackButton;
+            return redirectUrl;
+          }
+
+          if (appData.update?.review_beneficial_owners_individual?.length){
+            redirectUrl = reviewAllBeneficialOwnwer(appData, AllBoTypes.boiReview, AllBoTypes.boIndividual, beneficialOwnerIndividualReviewRedirectUrl);
+            return redirectUrl;
+          }
+          break;
+        }
+        case AllBoTypes.booReview: {
+          const booFromBackButton = checkForBackButtonBo(appData, AllBoTypes.boOther, beneficialOwnerOtherReviewRedirectUrl);
+          if (booFromBackButton) {
+            redirectUrl = booFromBackButton;
+            return redirectUrl;
+          }
+
+          if (appData.update?.review_beneficial_owners_corporate?.length){
+            redirectUrl = reviewAllBeneficialOwnwer(appData, AllBoTypes.booReview, AllBoTypes.boOther, beneficialOwnerOtherReviewRedirectUrl);
+            return redirectUrl;
+          }
+          break;
+        }
+    }
+  }
+  return redirectUrl;
+};
+
+const reviewAllBeneficialOwnwer = (appData: ApplicationData, boReviewType: string, boType: string, beneficialOwnerRedirectUrl: string) => {
+  let redirectUrl = "";
+  const boLength: number = appData[boType]?.length || 0;
+
+  if (boLength >= 0){
+    const bo = appData?.update?.[boReviewType]?.pop() as typeof boType;
+    if (!bo){
+      return redirectUrl;
+    }
+
+    let index = 0;
+    if (!appData[boType]) {
+      appData[boType] = [bo];
+    } else {
+      index = appData[boType].push(bo) - 1;
+    }
+    redirectUrl = `${beneficialOwnerRedirectUrl}${index}`;
+    return redirectUrl;
+  }
+  return redirectUrl;
```

Comment:
seems unsafe to return "". Is that checked for by the caller?

Reply:
to be addressed in Olayode's #869